### PR TITLE
Pagination in API, Identity Server and CLI

### LINKED ISSUE #453: Add pagination to list calls of the CLI
**Summary:** 

Add `--limit` and `--page` flags to list calls in the CLI.

**Why do we need this?**

In order to paginate the results, instead of using the defaults 0-10.

**What is already there? What do you see now?**

The RPCs already have it.

**What is missing? What do you want to see?**

Just the flags.

**How do you propose to implement this?**

Straightforward

**Can you do this yourself and submit a Pull Request?**

Yes, will do that asap.

### LINKED ISSUE #454: Make all List RPCs take ListXXX message as input
**Summary:** 

Make all List RPCs take ListXXX message as input instead of Identifiers.

Related to #422 / #423 

**Why do we need this?**

We may want to add pagination in the future. 

**What is already there? What do you see now?**

- `rpc ListRights(XxxIdentifiers) returns (Rights)`
- `rpc ListAPIKeys(XxxIdentifiers) returns (APIKeys)`
- `rpc List(google.protobuf.Empty) returns (Invitations)`

**What is missing? What do you want to see?**

These RPCs need to take a ListXxxRequest message instead of identifiers or empty.

**How do you propose to implement this?**

1. Update the API definitions
2. Regenerate the protos
3. Update the backend code
4. Update the JS SDK and console code

**Can you do this yourself and submit a Pull Request?**

I'll do 1-3. I'll need @kschiffer or @bafonins to do 4.

### PR DESCRIPTION
**Pull request by htdvisser**
**Summary:**

This PR closes #453 by adding the `--limit` and `--page` flags to the CLI. It also closes #454 by changing the API to have all `ListXXX` RPCs take a `ListXXXRequest` message with pagination fields. The Identity Server is updated to handle these in a consistent way.

**Changes:**

- Make all `ListXXX` RPCs take a `ListXXXRequest` message
- Make sure all `ListXXXRequest` messages have a `limit` and `page`
- Support pagination in all Identity Server List RPCs
- Add `--limit` and `--page` flags to the CLI's list commands, and make all of them forward the flags to the RPC.

**Notes for Reviewers:**

- @johanstokking: what's still left is the `ListApplicationWebhooks` RPC. I think we can add pagination there later if we want that.
- @johanstokking: do we want to add explicit validation rules for a maximum `limit` or should we make the maximum page size configurable (in the future)?

**Release Notes: _(optional)_**

- All `ListXXX` RPCs now take a `ListXXXRequest` message containing a `limit` and `page` for pagination.
- Introduced `--limit` and `--page` flags to all list commands of the CLI

### CODE REVIEW

File: cmd/ttn-lw-cli/commands/pagination.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	"strconv"
+
+	"github.com/spf13/pflag"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+func paginationFlags() *pflag.FlagSet {
+	flagSet := &pflag.FlagSet{}
+	flagSet.Uint32("limit", 10, "maximum number of results to get")
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/pagination.go]: This is a very low default

File: cmd/ttn-lw-cli/commands/pagination.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	limit, _ = flagSet.GetUint32("limit")
+	page, _ = flagSet.GetUint32("page")
+	responseHeaders := metadata.MD{}
+	opt = grpc.Header(&responseHeaders)
+	getTotal = func() uint64 {
+		totalHeader := responseHeaders.Get("x-total-count")
+		if len(totalHeader) > 0 {
+			total, _ := strconv.ParseUint(totalHeader[len(totalHeader)-1], 10, 64)
+			if total != 0 && total > uint64(limit)*uint64(page) {
+				logger.WithField("total", total).Info("Use pagination to get more results")
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/pagination.go]: Hint to `limit` and `page` flags?

File: pkg/identityserver/store/pagination.go [hunk 3]
```diff
@@ -16,44 +16,62 @@ package store
...
+		limit = uint32(md.Limit)
 	}
-	return md.Limit, offset
+	if page == 0 && md.Page != 0 {
+		page = uint32(md.Page)
+	}
+	if page == 0 {
+		page = 1
+	}
+	fmt.Println("WILL PAGINATE:", limit, page, (page-1)*limit)
```
Comment (johanstokking) [inline pkg/identityserver/store/pagination.go]: Remove

**Comment by johanstokking** [comment]
> * @johanstokking: what's still left is the `ListApplicationWebhooks` RPC. I think we can add pagination there later if we want that.

Agreed

> * @johanstokking: do we want to add explicit validation rules for a maximum `limit` or should we make the maximum page size configurable (in the future)?

Let's add a sane high limit in the API, like 1000 or 10000. Making it configurable later is nice, but having a maximum that's defined in the protos is a good signal and nice to have in all deployment models. We don't want things to blow up in a shared infrastructure environment as we don't support streaming results.

**Comment by kschiffer** [comment]
Applied an compatibility update to the JS SDK.

**Pull request merged**